### PR TITLE
Add check for unique file names as part of component

### PR DESCRIPTION
**Pull request by Svilen-Stefanov**
### **User description**
A friend ran an analysis and it resulted in quite a lot of file duplications within components. Adding a quick attempt to fix it here.

___

### **PR Type**
Bug fix, Tests

___

### **Description**
- Add deduplication of file assignments within components

- Implement `_ensure_unique_file_assignments()` method to remove duplicate files

- Preserve files across different components while removing duplicates within each

- Add comprehensive test coverage for file deduplication logic

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Component with duplicate files"] -->|_ensure_unique_file_assignments| B["Deduplicated file list"]
  C["Multiple components sharing files"] -->|Preserve across components| D["Files kept in all components"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>abstraction_agent.py</strong><dd><code>Add file deduplication step to pipeline</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents/abstraction_agent.py

<ul><li>Add call to <code>_ensure_unique_file_assignments()</code> as Step 7 in the <br>analysis pipeline<br> <li> Executes after ensuring unique key entities across components</ul>

</details>

  </td>
  <td><a href="https://github.com/CodeBoarding/CodeBoarding/pull/166/files#diff-38abcc624d53984ad24e385c9f0d6e109b3debf3ef4a5b4f502e6697ef7c026f">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>cluster_methods_mixin.py</strong><dd><code>Implement file deduplication within components</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents/cluster_methods_mixin.py

<ul><li>Implement <code>_ensure_unique_file_assignments()</code> method to deduplicate <br>files within each component<br> <li> Iterate through each component's <code>assigned_files</code> and remove duplicates <br>while preserving order<br> <li> Log debug messages for each removed duplicate and summary info for <br>total removed<br> <li> Allow same file to exist in multiple components but not within a <br>single component</ul>

</details>

  </td>
  <td><a href="https://github.com/CodeBoarding/CodeBoarding/pull/166/files#diff-57ed52582f09b181f5bb87ca7f5e7a0cebb8efa249d0cf0b64fdacd553b736fc">+29/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_cluster_methods_mixin.py</strong><dd><code>Add comprehensive tests for file deduplication</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/agents/test_cluster_methods_mixin.py

<ul><li>Add import for <code>AnalysisInsights</code> class<br> <li> Create new test class <code>TestEnsureUniqueFileAssignments</code> with helper <br>methods<br> <li> Add test for deduplication within single component with duplicate <br>files<br> <li> Add test for preservation of shared files across multiple components<br> <li> Add complex scenario test with multiple components and varied <br>deduplication needs</ul>

</details>

  </td>
  <td><a href="https://github.com/CodeBoarding/CodeBoarding/pull/166/files#diff-288b7802b9e33488d6a8325a9d2ce7e623ce2874449a7781bd7b489454ef5ea1">+67/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: agents/validation.py [hunk 1]
```diff
@@ -175,10 +176,11 @@ def _normalize_path(path: str) -> str:
 
     # Check 2: Are all component names valid?
     if context.valid_component_names:
-        invalid_classifications = []
-        for fc in result.file_paths:
-            if fc.component_name not in context.valid_component_names:
-                invalid_classifications.append(f"{fc.file_path} -> {fc.component_name}")
+        invalid_classifications = [
+            f"{fc.file_path} -> {fc.component_name}"
```
Comment (brovatten) [inline agents/validation.py]: Feel like previous is easier to read

File: agents/validation.py [hunk 2]
```diff
@@ -190,6 +192,22 @@ def _normalize_path(path: str) -> str:
...
+        for fc in result.file_paths
+        if file_counts[_normalize_path(fc.file_path)] > 1
+    ]
+
+    if duplicate_entries:
+        dup_str = ", ".join(duplicate_entries[:10])
+        more_msg = f" and {len(duplicate_entries) - 10} more" if len(duplicate_entries) > 10 else ""
+        feedback_messages.append(
+            f"Duplicate file assignments found: {dup_str}{more_msg}. "
+            f"Each file should be assigned to exactly one component."
```
Comment (brovatten) [inline agents/validation.py]: Hmmm... have we discussed this new assumption? Maybe we can minimize this overlap, but I think we all agreed there should be overlap sometimes.

Comment (Svilen-Stefanov) [reply agents/validation.py]: Good point. What I actually meant to do here is something different - guarantee that a file is present only one in a component because currently we can have a single file assigned many times to the same component. But you're right, the code is not doing what I intended.

File: agents/validation.py [hunk 3]
```diff
@@ -190,6 +196,31 @@ def _normalize_path(path: str) -> str:
...
+        else:
+            files_by_component[component].add(normalized)
+
+    if duplicate_entries:
+        dup_str = ", ".join(duplicate_entries[:10])
+        more_msg = f" and {len(duplicate_entries) - 10} more" if len(duplicate_entries) > 10 else ""
+        feedback_messages.append(
+            f"Duplicate file assignments found: {dup_str}{more_msg}. "
+            f"Each file should appear only once within a component."
+        )
```
Comment (ivanmilevtues) [inline agents/validation.py]: I think there is no need to have this as a validation step.
We can just remove the duplicates leave only one assigned file, no need to bother the LLM with this.

Add this logic as last step after fix_source_code_reference_lines (I suspect it might be it which introduces duplicates). Then whenever there is a duplicate remove it, simple fast no LLMs

Comment (brovatten) [reply agents/validation.py]: Yup agree. And we should get to the bottom of why the LLM would duplicate it to begin it

**Review by brovatten — APPROVED** [review]
I assume you mean one component had multiple of the same reference? If so, good job :D

**Review by ivanmilevtues — APPROVED** [review]
looks good now thanks/
